### PR TITLE
Link editing: Account for link anchor no longer being present when generating unique link instance key

### PR DESCRIPTION
**Pull request by getdave**
## Description
In https://github.com/WordPress/gutenberg/pull/34742 we used the unique anchor reference of each link in rich text to force remount the Link UI when moving between links (see that PR for why this was done).

Unfortunately, I forgot to account for instances where the anchor has been removed. This caused a `null` value to be set as the key of a `WeakMap` thus triggering a JS error.

In such cases we should not attempt to generate a unique link key and instead should bale out of the function. This causes the `key` prop on `<LinkControl>` to be `null` and thus not be set.

## How has this been tested?

* Create post.
* Create text.
* Create a link.
* Use arrows to select the link.
* Hit delete.
* You should see no console errors. 
* You should see no block errors.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/format-library/src/link/use-link-instance-key.js [hunk 1]
```diff
@@ -17,6 +17,9 @@ function getKey( _id ) {
  * @return {string} the unique key to use for this link control.
  */
 function useLinkInstanceKey( instance ) {
+	if ( ! instance ) {
```
Comment (getdave) [inline packages/format-library/src/link/use-link-instance-key.js:20]: This can be `null` or a `anchorRef`. So we need to avoid the null case.

**Review by andrewserong — APPROVED** [review]
Thanks for fixing this up! I could reproduce the issue, and confirm that this PR resolves it. Let's merge this in so we can cherry pick it into the 11.9 release.

**Comment by andrewserong** [comment]
I've just given this a rebase to get the tests passing.

File: packages/format-library/src/link/use-link-instance-key.js (hunk 1 above)
Comment (kevin940726) [reply packages/format-library/src/link/use-link-instance-key.js:20]: Could we add a comment above to explicitly state that we're returning a `undefined` here on purpose?

Comment (andrewserong) [reply packages/format-library/src/link/use-link-instance-key.js:20]: Apologies @kevin940726, I just missed your comment here before I hit "merge". I was perhaps over-eager to get this fix in, in time for the 11.9 release!

Comment (kevin940726) [reply packages/format-library/src/link/use-link-instance-key.js:20]: No worries 👍 , it's more like a nitpick anyway :).

**Comment by andrewserong** [comment]
Cherry picked into the Gutenberg 11.9 release in: https://github.com/WordPress/gutenberg/commit/3c11b5e82ea8c815718ee1ca0e802b1e52c8355a

**Comment by getdave** [comment]
Thanks @andrewserong (and @kevin940726!) 👏

**Pull request merged**